astropy/io/fits/file.py

```diff
@@ -426,13 +457,16 @@ def _open_fileobj(self, fileobj, mode, overwrite):
             self._file = fileobj
         elif isfile(fileobj):
             self._file = fileobj_open(self.name, IO_FITS_MODES[mode])
-        else:
-            self._file = gzip.open(self.name, IO_FITS_MODES[mode])
 
-        if fmode == 'ab+':
-            # Return to the beginning of the file--in Python 3 when opening in
-            # append mode the file pointer is at the end of the file
+        # Attempt to determine if the file represented by the open file object
+        # is compressed
+        try:
+            magic = self._file.read(4)
```

Comment:
The seek afterwards is definitely necessary in order to restore the file pointer after attempting to read the magic values. I'm not sure about a seek before reading the first bytes. The only way it would matter is if someone had read the file before passing it to `fits.open`, in which case you could argue that they deserve the error they will get.

I actually think that handling an exception here is a good idea. If there is a problem, it will still cause an error later on because the FITS header will not appear to be valid. But it will be more helpful than receiving a mysterious `IOError` here which could potentially be harder for a user to debug.